### ISSUE TITLE
CML AMIs

### ISSUE THREAD

**Issue by DavidGOrtega** [iterative/cml#162]
We are right now based on docker, however for cloud runners we could also work with AMIs. 
These images will remove the following pain points:

 - Accelerate the bootstrapping of the machine if it has properly installed the GPU drivers etc...
 - Not having to use  connections to hub.docker
 - For spot instances we could access the dispose mechanism easier

**Comment by andronovhopf** [comment]
First point seems like a major potential optimization. Pushing/pulling big docker images and installing libraries (for deep learning especially) every time seems to be a significant bottleneck on the [cloud-gpu-case](https://github.com/iterative/cml_cloud_case)- check out the timestamps in the Action logs.

 An AMI that requires environment configuration only once is potentially a huge time saver in the long run of a project.

**Comment by courentin** [comment]
That's a very good idea, do you have any ideas of an implementation ?
I'd like to help !

**Comment by DavidGOrtega** [comment]
@courentin thanks for your offering!
We are actually articulating this. Definitely the idea is generating an image with nvidia drivers and the Dockerfile contents with [packer](packer.io). The missing part would be adding a small bin to start the runners.

@courentin  If you have experience with packer, lets do it!

**Comment by DavidGOrtega** [comment]
This is now a real thing with the upcoming release of CML 0.3.0 AWS CML amis are available based on ubuntu 18.04

**Comment by 0x2b3bfa0** [comment]
> Currently we are recreating from scratch the instance environment each time we run a continuous integration job on Azure, because [(apparently)](https://docs.microsoft.com/en-us/azure/marketplace/marketplace-virtual-machines) we can't share instances images with a wider audience through the Azure Marketplace without having the [Microsoft Partner](https://docs.microsoft.com/en-us/partner-center/mpn-create-a-partner-center-account) qualification. (?)
> 
> The current workaround implies significantly longer total job execution times due to the increased instance configuration times, so being able to publish prebuilt CML base images to the Azure Marketplace —as we're already doing for AWS machine images— could be a nice improvement for end users.

**Comment by 0x2b3bfa0** [comment]
I was wondering if we could move [terraform-provider-iterative/cml](https://github.com/iterative/terraform-provider-iterative/tree/master/cml) to this repository. It would imply decoupling the region list from the provider code, but if we check it dynamically to solve https://github.com/iterative/terraform-provider-iterative/issues/99, it wouldn't matter anymore.

**Comment by 0x2b3bfa0** [comment]
Closing in favor of https://github.com/iterative/terraform-provider-iterative/issues/196
